src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,185 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+
+#include "devicetree.h"
+#include "util.h"
+
+#define FDT_V1_SIZE (7*4)
+
+#ifndef DEVICE_TREE_GUID
+#define DEVICE_TREE_GUID \
+        { 0xb1b621d5, 0xf19c, 0x41a5, {0x83, 0x0b, 0xd9, 0x15, 0x2c, 0x69, 0xaa, 0xe0} }
+#endif
+static const EFI_GUID fdt_guid = DEVICE_TREE_GUID;
+
+#ifndef EFI_DT_FIXUP_PROTOCOL_GUID
+#define EFI_DT_FIXUP_PROTOCOL_GUID \
+        { 0xe617d64c, 0xfe08, 0x46da, {0xf4, 0xdc, 0xbb, 0xd5, 0x87, 0x0c, 0x73, 0x00} }
+#endif
+static const EFI_GUID dt_fixup_guid = EFI_DT_FIXUP_PROTOCOL_GUID;
+
+#define EFI_DT_FIXUP_PROTOCOL_REVISION 0x00010000
+
+/* Add nodes and update properties */
+#define EFI_DT_APPLY_FIXUPS    0x00000001
+/*
+ * Reserve memory according to the /reserved-memory node
+ * and the memory reservation block
+ */
+#define EFI_DT_RESERVE_MEMORY  0x00000002
+
+typedef struct _EFI_DT_FIXUP_PROTOCOL EFI_DT_FIXUP_PROTOCOL;
+
+typedef EFI_STATUS (EFIAPI *EFI_DT_FIXUP) (
+        IN EFI_DT_FIXUP_PROTOCOL *This,
+        IN VOID                  *Fdt,
+        IN OUT UINTN             *BufferSize,
+        IN UINT32                Flags);
+
+struct _EFI_DT_FIXUP_PROTOCOL {
+        UINT64         Revision;
+        EFI_DT_FIXUP   Fixup;
+};
+
+static EFI_STATUS devicetree_get(struct devicetree_state *state) {
+        EFI_CONFIGURATION_TABLE *entry = ST->ConfigurationTable;
+        EFI_CONFIGURATION_TABLE *end = entry + ST->NumberOfTableEntries;
+
+        for (; entry < end; entry++) {
+                if (CompareGuid(&entry->VendorGuid, (EFI_GUID *)&fdt_guid) == 0) {
+                        state->orig = entry->VendorTable;
+                        return EFI_SUCCESS;
+                }
+        }
+
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_STATUS devicetree_allocate(struct devicetree_state *state, UINTN size) {
+        EFI_PHYSICAL_ADDRESS addr;
+        UINTN pages = DIV_ROUND_UP(size, EFI_PAGE_SIZE);
+        EFI_STATUS err;
+
+        err = uefi_call_wrapper(BS->AllocatePages, 4, AllocateAnyPages,
+                        EfiACPIReclaimMemory, pages, &addr);
+        if (EFI_ERROR(err)) {
+                Print(L"Error allocating device tree pages: %r\n", err);
+                return err;
+        }
+
+        state->addr = addr;
+        state->pages = pages;
+        return err;
+}
+
+static UINTN devicetree_allocated(struct devicetree_state *state) {
+        return state->pages * EFI_PAGE_SIZE;
+}
+
+static EFI_STATUS devicetree_free(struct devicetree_state *state) {
+        return uefi_call_wrapper(BS->FreePages, 2, state->addr, state->pages);
+}
+
+static VOID *devicetree_ptr(struct devicetree_state *state) {
+        return (VOID *)(uintptr_t)state->addr;
+}
+
+static EFI_STATUS devicetree_fixup(struct devicetree_state *state, UINTN len) {
+        EFI_DT_FIXUP_PROTOCOL *fixup;
+        UINTN size;
+        EFI_STATUS err;
+
+        err = LibLocateProtocol((EFI_GUID *)&dt_fixup_guid, (VOID **)&fixup);
```

Comment:
We can do without the global by using `&(EFI_GUID)EFI_DT_FIXUP_PROTOCOL_GUID` if I got the syntax right. (And the same for the other GUID)

Reply:
This will cost a little extra resources since every time `&(EFI_GUID)GUID` is used it'll need to be copied from read-only memory to the stack, whereas before we just casted the pointer to read-only memory and passed that. If you think it's worth the extra resources I'm totally fine with that.

Reply:
It's not like this is a particular hot path

Reply:
Nothing in sd-boot is a hot path. I meant the memory use. But as I said I'm fine with it, just wanted to make it clear that it is a tradeoff

Reply:
They end up on the stack. (it's c99 compund literals).